expression/collation.go

```diff
@@ -190,15 +190,15 @@ func deriveCollation(ctx sessionctx.Context, funcName string, args []Expression,
 	switch funcName {
 	case ast.Concat, ast.ConcatWS, ast.Lower, ast.Lcase, ast.Reverse, ast.Upper, ast.Ucase, ast.Quote, ast.Coalesce, ast.Greatest, ast.Least:
 		return CheckAndDeriveCollationFromExprs(ctx, funcName, retType, args...)
-	case ast.Left, ast.Right, ast.Repeat, ast.Trim, ast.LTrim, ast.RTrim, ast.Substr, ast.SubstringIndex, ast.Replace, ast.Substring, ast.Mid, ast.Translate:
+	case ast.Left, ast.Right, ast.Repeat, ast.Trim, ast.LTrim, ast.RTrim, ast.Substr, ast.SubstringIndex, ast.Replace, ast.Substring, ast.Mid, ast.Translate, ast.Regexp, ast.RegexpLike, ast.RegexpSubstr, ast.RegexpInStr, ast.RegexpReplace:
```

Comment:
In my understanding, `CheckAndDeriveCollationFromExprs` should at least take all string argument into account, @zimulala could you please give some explainations about this?

Reply:
no, from MySQL8.0 source code, some functions only check part of its arguments.
e.g. `select substring_index('aaa' collate utf8mb4_unicode_ci, 'b' collate utf8mb4_general_ci, 1);` will report error in MySQL5.7 but it is ok in MySQL8.0.26